crypto3/libs/blueprint/include/nil/blueprint/zkevm_bbf/opcodes/exp.hpp

```diff
@@ -53,81 +54,119 @@ namespace nil {
+
+                        allocate(D_gt_diff[i], 16 + i, 1);
+                        allocate(D_gt_255[i], i, 0);
+                        constrain(D_gt_255[i] * (1 - D_gt_255[i]));
+                        constrain(D[i] + D_gt_diff[i] - 0xFF - D_gt_255[i] * 0x10000);
+
+                        allocate(D_first_inv[i], 32 + i, 1);
+                        // D_first_inv[i] != 0 => it is an inverse
+                        constrain(D_first_inv[i] * (1 - D[i] * D_first_inv[i]));
+
+                        d_sum += D[i];
+                        d_first_nonzero_sum += D[i] * D_first_inv[i];
                     }
 
-                    //  (s == 0 && d == 0) => R == 1
-                    constrain((s - 1) * (d_sum - 1) * (R[15] - 1));
-                    constrain((s - 1) * (d_sum - 1) * (r_sum - R[15]));
 
                     auto A_128 = chunks16_to_chunks128<TYPE>(A);
                     auto D_128 = chunks16_to_chunks128<TYPE>(D);
                     auto R_128 = chunks16_to_chunks128<TYPE>(R);
-                    constrain( (1 - (d_sum - 1) * d_sum_inv_1) * (A_128.first - R_128.first) );
-                    constrain( (1 - (d_sum - 1) * d_sum_inv_1) * (A_128.second - R_128.second) );
-                    constrain( (1 - (d_sum - 1) * d_sum_inv_1) * s );
 
+                    // If d[i] != 0, for j > i D_first_inv[j] = 0, since it isn't first
+                    for (std::size_t i = 0; i < 15; ++i) {
+                        TYPE next_invs;
+                        for (int j = i + 1; j < 16; ++j) next_invs += D[j] * D_first_inv[j];
+                        constrain(D[i] * next_invs);
+                    }
+
+                    // If d != 0, d_first_inv has a non-zero element
+                    constrain(d_sum * (1 - d_first_nonzero_sum));
+                    // else, the sum is 0 so it's a [D != 0] too
+                    auto &d_ne_0 = d_first_nonzero_sum;
+
+                    TYPE d_len;
+                    for (std::size_t i = 0; i < 16; ++i) {
+                        d_len += D[i] * D_first_inv[i] * (
+                            D_gt_255[i] + 1 // this word's bytes
+                            + 2 * (15 - i)  // less significant bytes
+                        );
+                    }
+                    allocate(d_len, 32, 2);
+
+                    // D = 1 <=> d[15] is first non-zero and d[15] = 1
+                    TYPE d_is_1_w;
+                    if constexpr (stage == GenerationStage::ASSIGNMENT)
+                        d_is_1_w = D[15] == 1 ? 0 : (D[15] - 1).inversed();
+                    allocate(d_is_1_w, 33, 2);
+                    constrain((D[15] - 1) * (1 - (D[15] - 1) * d_is_1_w));
+
+                    TYPE d_is_1 = (1 - (D[15] - 1) * d_is_1_w) * D[15] * D_first_inv[15];
+
+                    allocate(s, 34, 2);
+                    constrain(s * (s-1));
+
+                    // s == 0 => d == 0 || d == 1
+                    constrain((1 - s) * d_ne_0 * (1 - d_is_1));
```

Comment:
Too high degree d_is_1 is not assigned, so it has degree 4.
Final opcode constraint has degree 6